platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,518 @@
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     * @see #resumeCompaction()
+     */
+    public void pauseCompaction() throws IOException {
+        snapshotCompactionLock.acquireUninterruptibly();
+        // Check if compaction is currently in progress. If so, flush and close the current file, so
+        // it's included to the snapshot
+        final DataFileWriter<?> compactionWriter = currentWriter.get();
+        if (compactionWriter != null) {
+            compactionWasInProgress.set(true);
+            compactionLevelInProgress.set(compactionWriter.getMetadata().getCompactionLevel());
+            finishCurrentCompactionFile();
+            // Don't start a new compaction file here, as it would be included to snapshots, but
+            // it shouldn't, as it isn't fully written yet. Instead, a new file will be started
+            // right after snapshot is taken, in resumeCompaction()
+        }
+        // Don't release the lock here, it will be done later in resumeCompaction(). If there is no
+        // compaction currently running, the lock will prevent starting a new one until snapshot is
+        // done
+    }
+
+    /**
+     * Resumes compaction previously put on hold with {@link #pauseCompaction()}. If there was no
+     * compaction running at that moment, but new compaction was started (and blocked) since {@link
+     * #pauseCompaction()}, this new compaction is resumed.
+     * <p>
+     * <b>This method must be always balanced with and called after {@link #pauseCompaction()}. If
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    public void resumeCompaction() throws IOException {
+        try {
+            if (compactionWasInProgress.getAndSet(false)) {
+                assert currentWriter.get() == null;
+                assert currentReader.get() == null;
+                startNewCompactionFile(compactionLevelInProgress.getAndSet(0));
+            }
+        } finally {
+            snapshotCompactionLock.release();
+        }
+    }
+
+    /**
+     * Compact data files in the collection according to the compaction algorithm.
+     *
+     * @throws IOException          if there was a problem merging
+     * @throws InterruptedException if the merge thread was interrupted
+     * @return true if compaction was performed, false otherwise
+     */
+    @SuppressWarnings("unchecked")
+    public boolean compact() throws IOException, InterruptedException {
+
+        final List<? extends DataFileReader<?>> allCompactableFiles = dataFileCollection.getAllCompletedFiles();
+        final List<DataFileReader<?>> filesToCompact = compactionPlan(
```

Comment:
Although compactionPlan() can work with any data file reader type, it may be cleaner to change it to accept `DataFileReader<D>` rather than `DataFileReader<?>`. It would eliminate type casts here and would get rid of suppressed "unchecked" warnings.

Reply:
`DataFileCompactor` is intentionally not parametrized as it's not working with the elements of `DataFileReader` and it's not supposed to know the underlying type. However, I see your point - the warning suppression and explicit casts don't look good. So I changed the type to get rid of both.